src/c2goto/library/stdlib.c

```diff
@@ -265,6 +266,51 @@ void kfree(const void *ptr)
 {
   free((void *)ptr);
 }
+unsigned long copy_to_user(void *to, void *from, unsigned size)
+{
+  //checking on the passed parameters of kernel function
+  //the source in kernel space and destination in user space must be valid
+  assert(to != NULL);
+  assert(from != NULL);
+  assert(size <= PAGE_SIZE);
+
+  //check on user space memory boundry
+  assert(
+    (char *)to < user_memory + USER_MEMORY_SPACE && (char *)to >= user_memory);
+  //check on kernel space memory boundry
+  assert(
+    (char *)from < kernel_memory + KERNEL_MEMORY_SPACE &&
+    (char *)from >= kernel_memory);
+  //copy memory from kernel space to user space
+  //simulate the copy operation by memcpy
+  memcpy(to, from, size);
+
+  return 0;
+}
+
+unsigned long copy_from_user(void *to, void *from, unsigned size)
+{
+  //the source in user space and destination in kernel space must be valid
+  //avoid dereferencing null pointer
+  assert(to != NULL);
+  assert(from != NULL);
+  assert(size <= PAGE_SIZE);
+
+  //check on kernel space memory boundry
+  assert(
+    (char *)to < kernel_memory + KERNEL_MEMORY_SPACE &&
+    (char *)to >= kernel_memory);
+  //check on user space memory boundry
+  assert(
+    (char *)from < user_memory + USER_MEMORY_SPACE &&
+    (char *)from >= user_memory);
```

Comment:
Why are these asserts needed? ESBMC should be able to find array-out-of-bounds without them. Same for `copy_to_user()`.

Reply:
these assertions are used for verifying if the simulation would cause the copying out of the available/legal space. Yes so just as you said, I could let ESBMC directly find the array-out-of-bounds problem ?

Reply:
> Yes so just as you said, I could let ESBMC directly find the array-out-of-bounds problem ?

Sounds good. Will you get the same results?